lib/helpers.js

```diff
@@ -13,71 +23,48 @@ var assert = require('assert-plus');
+ * 1) new HttpError(sprintf_args...);
+ * 2) new HttpError(anotherErr, sprintf_args);
+ * 3) new HttpError({...}, sprintf_args);
+ * restify-errors aims only value is to add support for additional options
+ * using the signature #3. this function parses out the arguments specific to
+ * restify-errors so that they don't get passed on to VError.
  * @public
- * @function parseVariadicArgs
- * @param    {Object}  ctorArgs    the passed in arguments object to error
- * constructor
- * @param    {Boolean} werrorSuper if the super being invoked is going to
- * WError
- * @returns  {Object}
+ * @param {Object} ctorArgs an 'arguments' object
+ * @function parseVErrorArgs
+ * @returns {Object}
  */
-function parseVariadicArgs(ctorArgs, werrorSuper) {
-
-    function parse() {
-
-        // to array the inner arguments
-        var args = _.toArray(arguments);
-        var options = null;
-
-        // 1)
-        if (arguments[0] instanceof Error) {
-            // 1)
-            if (_.isPlainObject(arguments[1])) {
-                options = arguments[1] || {};
-
-                // if we're calling WError super, don't pass in the special
-                // restify options object.
-                if (werrorSuper === true) {
-                    args = [arguments[0]];
-                }
-            }
-            // 4)
-            // otherwise, pass through to WError
-        }
-        // 2)
-        else if (_.isPlainObject(arguments[0])) {
-            options = arguments[0] || {};
-
-            // if we're calling WError super, don't pass in the special restify
-            // options object. pass all subsequent args, which could include
-            // extsprintf style strings.
-            if (werrorSuper === true) {
-                // drop first arg
-                args = _.drop(args);
+function parseVErrorArgs(ctorArgs) {
+    // to array the inner arguments so it's easier to determine which cases
+    // we are looking at
+    var args = _.toArray(ctorArgs);
+    var internalOpts = {};
+    var verrorOpts = {};
+    var verrorArgs;
+
+    if (_.isPlainObject(args[0])) {
+        // split restify-errors options from verror options
+        _.forOwn(args[0], function(val, key) {
+            if (INTERNAL_OPTS_KEYS.hasOwnProperty(key)) {
+                internalOpts[key] = val;
+            } else {
+                verrorOpts[key] = val;
             }
+        });
 
-            // if we have an options object, assert them.
-            assert.optionalObject(options, 'options');
-            assert.optionalString(options.message, 'options.message');
-            assert.optionalNumber(options.statusCode, 'options.statusCode');
-        }
-        // 3
-        // otherwise, pass through to WError
-
-        return {
-            args: args,
-            options: options
-        };
+        // reconstruct verror ctor options from the cleaned up options
+        verrorArgs = [ verrorOpts ].concat(_.tail(args));
```

Comment:
Why is `verrorArgs` not just `verrorOpts`? Doesn't adding `args` back to `verrorOpts` result in the overlap of the sets being included in the array twice?

Reply:
VError has various signatures, the one we care about is `new VError({}, 'sprintf', 'sprintf args' ...)`

restify-errors options can be appended to the first arg - so what we do here is strip out restify-error specific options from `opts`, rename the options leftover as `verrorOpts`, then concat with everything but the first argument of raw `args` (the first argument being itself).